include/zephyr/sys/util.h

```diff
@@ -110,6 +110,45 @@ extern "C" {
 
 #endif /* __cplusplus */
 
+/**
+ * @brief Whether @p ptr is an element of @p array
+ *
+ * This macro can be seen as a slightly stricter version of @ref PART_OF_ARRAY
+ * in that it also ensures that @p ptr is aligned to an array-element boundary
+ * of @p array.
+ *
+ * In C, passing a pointer as @p array causes a compile error.
+ *
+ * @param array the array in question
+ * @param ptr the pointer to check
+ *
+ * @return 1 if @p ptr is part of @p array, 0 otherwise
+ */
+#define IS_ARRAY_ELEMENT(array, ptr)                                                               \
+	((uint8_t *)&(array)[0] <= (uint8_t *)(ptr) &&                                             \
+	 (uint8_t *)(ptr) < (uint8_t *)&(array)[ARRAY_SIZE(array)] &&                              \
+	 ((uint8_t *)(ptr) - (uint8_t *)(array)) % sizeof((array)[0]) == 0)
```

Comment:
What's the difference between this and `PART_OF_ARRAY`?

Reply:
The answer is actually right there in the doc.

Also, see tests.

In short, this returns 0, when a pointer is not aligned to an array-element boundary